File: crates/starscape/src/catalog.rs
//! Validation for the two fetched star assets.
//!
//! `bright.bin` is the GPU catalog: an 8-byte `STR1` header and then one
//! 20-byte record per star — three f32 of unit position, one f32 magnitude,
//! four bytes of RGBA colour — sorted brightest first so a partial stream is
//! still the brightest sky rather than a random subset of it.
//!
//! `named.json` is the annotation catalog: a small IAU/SIMBAD-attributed list
//! that points at records in `bright.bin` by index.

use serde::Deserialize;

const MAGIC: &[u8; 4] = b"STR1";
pub const HEADER_LEN: usize = 8;
pub const STRIDE: usize = 20;
const MAX_STARS: usize = 1_000_000;

/// Parse and check the header, returning the star count it declares.
pub fn header_count(bytes: &[u8]) -> Result<usize, &'static str> {
    if bytes.len() < HEADER_LEN || &bytes[..4] != MAGIC {
        return Err("bad star asset header");
    }
    let count = u32::from_le_bytes(bytes[4..8].try_into().map_err(|_| "bad star count")?) as usize;
    if count == 0 || count > MAX_STARS || count > i32::MAX as usize {
        return Err("star count out of bounds");
    }
    Ok(count)
}

/// Full structural validation: exact length, and every record a finite unit
/// vector. A star at the origin or with a NaN coordinate renders as a
/// full-screen artefact rather than as nothing, which is why this is checked
/// rather than trusted.
pub fn validate(bytes: &[u8]) -> Result<usize, &'static str> {
    let count = header_count(bytes)?;
    let total = HEADER_LEN
        .checked_add(
            count
                .checked_mul(STRIDE)
                .ok_or("star asset length overflow")?,
        )
        .ok_or("star asset length overflow")?;
    if bytes.len() != total {
        return Err("star asset length mismatch");
    }
    for record in bytes[HEADER_LEN..].chunks_exact(STRIDE) {
        let at = |offset: usize| {
            f32::from_le_bytes(record[offset..offset + 4].try_into().expect("four bytes"))
        };
        let (x, y, z, magnitude) = (at(0), at(4), at(8), at(12));
        let norm_squared = x * x + y * y + z * z;
        if !magnitude.is_finite() || !(0.98..=1.02).contains(&norm_squared) {
            return Err("invalid star record");
        }
    }
    Ok(count)
}

/// The J2000 unit vector of the star at `index`, read straight out of the
/// fetched bytes so the annotations need no second copy of the catalog.
#[must_use]
pub fn position(bytes: &[u8], index: usize) -> Option<[f64; 3]> {
    let offset = HEADER_LEN.checked_add(index.checked_mul(STRIDE)?)?;
    let record = bytes.get(offset..offset.checked_add(12)?)?;
    Some(std::array::from_fn(|axis| {
        f64::from(f32::from_le_bytes(
            record[axis * 4..axis * 4 + 4]
                .try_into()
                .expect("four bytes"),
        ))
    }))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NamedStar {
    pub bright_index: usize,
    pub name: String,
    pub constellation: String,
    pub classification: String,
    pub distance_ly: f64,
}

#[derive(Debug, Deserialize)]
pub struct NamedCatalog {
    pub version: u8,
    pub sources: Vec<String>,
    pub stars: Vec<NamedStar>,
}

impl NamedCatalog {
    pub fn parse(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const BRIGHT: &[u8] = include_bytes!("../../../static/stars/bright.bin");
    const NAMED: &str = include_str!("../../../static/stars/named.json");

    fn one_record(x: f32, y: f32, z: f32) -> Vec<u8> {
        let mut bytes = b"STR1\x01\0\0\0".to_vec();
        for value in [x, y, z, 2.0] {
            bytes.extend_from_slice(&value.to_le_bytes());
        }
        bytes.extend_from_slice(&[255, 255, 255, 255]);
        bytes
    }

    #[test]
    fn the_shipped_catalog_validates_and_has_its_published_count() {
        assert_eq!(validate(BRIGHT).expect("the shipped catalog"), 12_191);
    }

    #[test]
    fn the_catalog_is_sorted_brightest_first_so_a_partial_stream_is_still_the_sky() {
        let magnitudes: Vec<f32> = BRIGHT[HEADER_LEN..]
            .chunks_exact(STRIDE)
            .map(|record| f32::from_le_bytes(record[12..16].try_into().unwrap()))
            .collect();
        assert!(magnitudes.windows(2).all(|pair| pair[0] <= pair[1]));
    }

    #[test]
    fn a_truncated_or_absurd_header_is_refused() {
        assert_eq!(
            validate(b"STR1\0\0\0\0").err(),
            Some("star count out of bounds")
        );
        assert_eq!(
            validate(b"STR1\xff\xff\xff\xff").err(),
            Some("star count out of bounds")
        );
        assert_eq!(
            validate(b"NOPE\x01\0\0\0").err(),
            Some("bad star asset header")
        );
        let mut truncated = one_record(1.0, 0.0, 0.0);
        truncated.pop();
        assert_eq!(
            validate(&truncated).err(),
            Some("star asset length mismatch")
        );
    }

    #[test]
    fn a_record_that_is_not_a_finite_unit_vector_is_refused() {
        for bad in [
            one_record(0.0, 0.0, 0.0),
            one_record(f32::NAN, 0.0, 1.0),
            one_record(2.0, 0.0, 0.0),
        ] {
            assert_eq!(validate(&bad).err(), Some("invalid star record"));
        }
    }

    #[test]
    fn a_star_position_is_readable_by_index_and_bounded_by_the_catalog() {
        let vector = position(BRIGHT, 0).expect("the brightest star");
        let norm: f64 = vector.iter().map(|c| c * c).sum();
        assert!((norm - 1.0).abs() < 0.02, "{norm}");
        assert!(position(BRIGHT, 12_191).is_none());
    }

    #[test]
    fn the_named_catalog_is_structured_and_attributed_to_its_sources() {
        let catalog = NamedCatalog::parse(NAMED).expect("named.json");
        assert_eq!(catalog.version, 1);
        assert!(catalog.stars.len() >= 50);
        assert!(catalog.sources.iter().any(|source| source.contains("IAU")));
        assert!(
            catalog
                .sources
                .iter()
                .any(|source| source.contains("SIMBAD"))
        );
        for star in &catalog.stars {
            assert!(
                position(BRIGHT, star.bright_index).is_some(),
                "{}",
                star.name
            );
            assert!(!star.name.is_empty() && !star.constellation.is_empty());
            assert!(!star.classification.is_empty());
            assert!(star.distance_ly.is_finite() && star.distance_ly > 0.0);
        }
    }
}

File: src/features/sky/catalog.ts
/** The two fetched star assets, and the validation that stands between the
 * network and the canvas.
 *
 * `bright.bin` is the catalog: an 8-byte `STR2` header and then one 32-byte
 * record per star — three f32 of unit position, one f32 magnitude, four bytes
 * of RGBA colour, a u64 catalogue id, and the byte that says which catalogue
 * that id belongs to — sorted brightest first, so a partial read is still the
 * brightest sky rather than a random subset of it.
 *
 * The id is what makes a star addressable: everything past this leg (streamed
 * depth, picking, a detail panel that can look a star up) needs to say *which*
 * star, and an index into a file that is rebuilt whenever the catalogue is
 * cannot say that. `tools/starcat/build_bright.py` writes the format.
 *
 * `named.json` is the annotation catalog: a small IAU/SIMBAD-attributed list
 * that points at records in `bright.bin` by index — and the builder refuses to
 * publish a catalog whose 50 named stars do not all still match by position.
 */

import type { Vec3 } from './sidereal';

export const HEADER_LEN = 8;
export const STRIDE = 32;
const MAGIC = 0x32_52_54_53; // "STR2", little-endian.
const MAX_STARS = 1_000_000;

/** Which catalogue a record's id belongs to. */
export const KIND_GAIA = 0;
export const KIND_HIP = 1;

export class CatalogError extends Error {}

/** Structure of arrays, because the canvas walks all of it every frame. */
export interface StarCatalog {
  /** J2000 unit vectors, three per star. */
  position: Float32Array;
  magnitude: Float32Array;
  /** RGB in 0..1, three per star. */
  color: Float32Array;
  /** Gaia DR3 `source_id` or Hipparcos number, one per star. */
  id: BigUint64Array;
  /** {@link KIND_GAIA} or {@link KIND_HIP}, one per star. */
  kind: Uint8Array;
  count: number;
}

/** Parse and check the header, returning the star count it declares. */
export function headerCount(bytes: Uint8Array): number {
  if (bytes.byteLength < HEADER_LEN) throw new CatalogError('bad star asset header');
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (view.getUint32(0, true) !== MAGIC) throw new CatalogError('bad star asset header');
  const count = view.getUint32(4, true);
  if (count === 0 || count > MAX_STARS) throw new CatalogError('star count out of bounds');
  return count;
}

/** Full structural validation and unpacking: exact length, and every record a
 * finite unit vector. A star at the origin or with a NaN coordinate renders as
 * a full-screen artefact rather than as nothing, which is why this is checked
 * rather than trusted. */
export function parseStars(bytes: Uint8Array): StarCatalog {
  const count = headerCount(bytes);
  if (bytes.byteLength !== HEADER_LEN + count * STRIDE) {
    throw new CatalogError('star asset length mismatch');
  }
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const position = new Float32Array(count * 3);
  const magnitude = new Float32Array(count);
  const color = new Float32Array(count * 3);
  const id = new BigUint64Array(count);
  const kind = new Uint8Array(count);

  for (let index = 0; index < count; index += 1) {
    const at = HEADER_LEN + index * STRIDE;
    const x = view.getFloat32(at, true);
    const y = view.getFloat32(at + 4, true);
    const z = view.getFloat32(at + 8, true);
    const mag = view.getFloat32(at + 12, true);
    const normSquared = x * x + y * y + z * z;
    if (!Number.isFinite(mag) || !(normSquared >= 0.98 && normSquared <= 1.02)) {
      throw new CatalogError('invalid star record');
    }
    position[index * 3] = x;
    position[index * 3 + 1] = y;
    position[index * 3 + 2] = z;
    magnitude[index] = mag;
    color[index * 3] = view.getUint8(at + 16) / 255;
    color[index * 3 + 1] = view.getUint8(at + 17) / 255;
    color[index * 3 + 2] = view.getUint8(at + 18) / 255;
    id[index] = view.getBigUint64(at + 20, true);
    const which = view.getUint8(at + 28);
    if (which !== KIND_GAIA && which !== KIND_HIP) {
      throw new CatalogError('invalid star record');
    }
    kind[index] = which;
  }
  return { position, magnitude, color, id, kind, count };
}

/** How a star names itself: the form `/api/sky/star/<key>` will read, and the
 * form a callout or a pick shows when the star has no proper name. */
export function starKey(catalog: StarCatalog, index: number): string | null {
  if (!Number.isInteger(index) || index < 0 || index >= catalog.count) return null;
  const prefix = catalog.kind[index] === KIND_HIP ? 'hip' : 'gaia';
  return `${prefix}-${catalog.id[index]!}`;
}

/** The J2000 unit vector of the star at `index`. */
export function starPosition(catalog: StarCatalog, index: number): Vec3 | null {
  if (!Number.isInteger(index) || index < 0 || index >= catalog.count) return null;
  return [
    catalog.position[index * 3]!,
    catalog.position[index * 3 + 1]!,
    catalog.position[index * 3 + 2]!,
  ];
}

export interface NamedStar {
  brightIndex: number;
  name: string;
  constellation: string;
  classification: string;
  distanceLy: number;
}

export interface NamedCatalog {
  version: number;
  sources: string[];
  stars: NamedStar[];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

/** Parse `named.json`. Nothing about a catalog that arrived over the network
 * is trusted: a name that is not a string would end up in the DOM, and an
 * index that is not one would label the wrong star. */
export function parseNamed(json: string): NamedCatalog {
  const parsed: unknown = JSON.parse(json);
  if (!isRecord(parsed) || typeof parsed.version !== 'number') {
    throw new CatalogError('named.json is not a catalog');
  }
  const sources = Array.isArray(parsed.sources)
    ? parsed.sources.filter((source): source is string => typeof source === 'string')
    : [];
  if (!Array.isArray(parsed.stars) || parsed.stars.length === 0) {
    throw new CatalogError('named.json has no stars');
  }
  const stars = parsed.stars.map((entry): NamedStar => {
    if (
      !isRecord(entry) ||
      !Number.isInteger(entry.brightIndex) ||
      typeof entry.name !== 'string' ||
      entry.name === '' ||
      typeof entry.constellation !== 'string' ||
      entry.constellation === '' ||
      typeof entry.classification !== 'string' ||
      entry.classification === '' ||
      typeof entry.distanceLy !== 'number' ||
      !Number.isFinite(entry.distanceLy) ||
      entry.distanceLy <= 0
    ) {
      throw new CatalogError('named.json holds a malformed star');
    }
    return {
      brightIndex: entry.brightIndex as number,
      name: entry.name,
      constellation: entry.constellation,
      classification: entry.classification,
      distanceLy: entry.distanceLy,
    };
  });
  return { version: parsed.version, sources, stars };
}

/** The J2000 vectors the named catalog points at. A catalog whose indices do
 * not all resolve is a mismatched pair of assets, and labelling the wrong
 * stars is worse than labelling none. */
export function namedVectors(catalog: StarCatalog, named: NamedCatalog): Vec3[] {
  return named.stars.map((star) => {
    const position = starPosition(catalog, star.brightIndex);
    if (!position) throw new CatalogError('named.json points outside the star catalog');
    return position;
  });
}
